ocp_resources/utils.py

```diff
@@ -25,76 +30,183 @@ class TimeoutSampler:
+
+        return output
+
     @property
     def _func_log(self):
         return f"Function: {self.func} Kwargs: {self.func_kwargs}"
 
     def __iter__(self):
-        last_exp = None
         timeout_watch = TimeoutWatch(timeout=self.wait_timeout)
         if self.print_log:
             LOGGER.info(
                 f"Waiting for {self.wait_timeout} seconds, retry every {self.sleep} seconds"
             )
 
+        last_exp_log = "No exceptions raised"
         while timeout_watch.remaining_time() > 0:
             try:
                 self.elapsed_time = self.wait_timeout - timeout_watch.remaining_time()
                 yield self.func(**self.func_kwargs)
                 self.elapsed_time = None
+
                 time.sleep(self.sleep)
 
-            except self.exception as exp:
-                last_exp = exp
-                self.elapsed_time = None
+            except self._exceptions as exp:
+                last_exp_log = self._get_exception_log(exp=exp)
+                if self._is_raisable_exception(exp=exp):
+                    LOGGER.error(last_exp_log)
+                    raise exp
+                else:
+                    LOGGER.warning(f"{last_exp_log}: Retrying")
 
+                self.elapsed_time = None
                 time.sleep(self.sleep)
 
             finally:
                 if self.elapsed_time and self.print_log:
                     LOGGER.info(f"Elapsed time: {self.elapsed_time}")
 
-        raise TimeoutExpiredError(self._process_execution(exp=last_exp))
+        raise TimeoutExpiredError(last_exp_log)
 
-    def _process_execution(self, exp=None):
-        exp_name = exp.__class__.__name__ if exp else "N/A"
-        last_exception_log = f"Last exception: {exp_name}: {exp}"
+    @staticmethod
+    def _is_exception_message_matched(exp, exception_messages):
+        if not exception_messages:
+            return True
 
-        log = "{timeout}\n{func_log}\n{last_exception_log}".format(
-            timeout=self.wait_timeout,
-            func_log=self._func_log,
-            last_exception_log=last_exception_log,
-        )
+        for msg in exception_messages:
+            if msg and msg in str(exp):
+                return True
+
+        return False
 
-        if self.exceptions_msg:
-            if self.exceptions_msg not in str(exp):
-                LOGGER.error(log)
-                raise exp
-            else:
-                LOGGER.warning(f"{self.exceptions_msg}: Retrying")
+    def _is_raisable_exception(self, exp):
+        """
+        Verify whether exception should be raised during execution
+
+        The raised exception will be EITHER:
+        An exception specifically referenced in exceptions_dict
+        OR
+        A child exception inherited from an entry in exceptions_dict
+
+        Args:
+            exp: exception: Exception object to be tested
+
+        Returns:
+            bool: True if exception should be raised, False otherwise
+        """
```

Comment:
I've updated the docstring. We can do this as a single `for entry in dict`, which would cover all cases, if we're willing to drop the `if exp_class in dict` block.

Reply:
https://github.com/RedHatQE/openshift-python-wrapper/commit/71246eb5557bc6bc797a34f24c1c73cd9e3bda6f